float32[] imgState
float64[] rState
